### PR TITLE
Normative: Add Optional Chaining

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This change adds the relevant updates for the stage 3 [Optional Chaining](https://github.com/tc39/proposal-optional-chaining/) proposal. The intended changes can be viewed at https://tc39.es/proposal-optional-chaining/.

The Test262 PR is available at https://github.com/tc39/test262/pull/2212

### CODE REVIEW

**Comment by ljharb** [comment]
@DanielRosenwasser please either link to the test262 tests PR, or put one up, prior to the December meeting. Thanks!

File: spec.html [hunk 1]
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+            OptionalExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
+          1. Let _baseReference_ be ? _baseExpression_.Evaluation().
+          1. Let _baseValue_ be ? GetValue(_baseReference_).
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
```
Comment (bakkot) [inline spec.html]: I think this is less clear than just having three different `Runtime Semantics: Evaluation ` sections, one for each of the three kinds of chain, each with their own copy of the steps preceding this line.

File: spec.html [hunk 2]
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+        <h1>Runtime Semantics: Evaluation</h1>
+        <emu-grammar>
+          OptionalExpression :
+            MemberExpression OptionalChain
+            CallExpression OptionalChain
+            OptionalExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
+          1. Let _baseReference_ be ? _baseExpression_.Evaluation().
```
Comment (bakkot) [inline spec.html]:
```suggestion
          1. Let _baseReference_ be the result of evaluating _baseExpression_.
```

(There's no need for the `?`; that's [handled by](https://tc39.es/ecma262/#sec-getvalue) the `GetValue` in subsequent step.)

File: spec.html (hunk 1 above)
Comment (DanielRosenwasser) [reply spec.html]: Because I spent a lot of time misunderstanding this, I should probably clarify before I do any more work 😄

Basically the line that you actually take issue with is this, right?

```
1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
```

It feels repetitive to have 3 different sections, but I can make the change.

Comment (bakkot) [reply spec.html]: I do also take some issue with that line, but actually it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records, not to dispatch abstract operations on parse nodes.

I agree it is a tiny bit repetitive, but I think the added clarity is worth it. And it's not actually that much code which gets repeated.

Comment (devsnek) [reply spec.html]: I also found this to be a bit weird. +1 to separate runtime semantics.

Comment (zenparsing) [reply spec.html]: I agree with @bakkot here. In general, I think it's best to stick to the "syntax-directed evaluation" style as much as possible.

Comment (bakkot) [reply spec.html]: To be explicit, what I think would be the most clear would be to have three subsections under the current `Runtime Semantics: Evaluation`, each in the mold of

```
        <emu-grammar>
          OptionalExpression :
            MemberExpression OptionalChain
        </emu-grammar>
        <emu-alg>
          1. Let _baseReference_ be the result of evaluating |MemberExpression|.
          1. Let _baseValue_ be ? GetValue(_baseReference_).
          1. If _baseValue_ is *undefined* or *null*, then
            1. Return *undefined*.
          1. Return the result of performing ChainEvaluation for |OptionalChain| with arguments _baseValue_ and _baseReference_.
        <emu-alg>
```

(And to similarly rewrite the uses of `ChainEvaluation` in the `Runtime Semantics: ChainEvaluation` section to use this style of dispatch.)

File: spec.html [hunk 3]
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+        </emu-alg>
+        <emu-grammar>OptionalChain : `?.` Arguments</emu-grammar>
+        <emu-alg>
+          1. Let _thisChain_ be this production.
+          1. Let _tailCall_ be IsInTailPosition(_thisChain_).
+          1. Return ? EvaluateCall(_baseValue_, _baseReference_, |Arguments|, _tailCall_).
+        </emu-alg>
+        <emu-grammar>OptionalChain : OptionalChain `[` Expression `]`</emu-grammar>
+        <emu-alg>
+          1. Let _optionalChain_ be this |OptionalChain|.
```
Comment (bakkot) [inline spec.html]:
```suggestion
          1. Let _optionalChain_ be |OptionalChain|.
```

The `this |Nonterminal|` wording refers to the the entire production, rather than one of the nonterminals on the RHS. Saying `Let _optionalChain_ be |OptionalChain|` would refer to the one on the RHS. See for example [the Evaluation semantics for](https://tc39.es/ecma262/#sec-function-calls-runtime-semantics-evaluation) `CallExpression : CallExpression Arguments`, which has exactly the same ambiguity, and which in step 1 uses a bare `CallExpression` to mean the one on the RHS and in step 3 uses `this CallExpression` to mean the full production.

Here this line only makes sense if it is intended to refer to the one one the RHS rather than the full production.

(Ditto most other uses of `this |NonTerminal|` in this PR.)

File: spec.html (hunk 1 above)
Comment (jmdyck) [reply spec.html]: > it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records,

And module records, and internal methods.

> not to dispatch abstract operations on parse nodes.

But see PR #955.

File: spec.html [hunk 4]
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+          1. Let _newReference_ be ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
+          1. Let _newValue_ be ? GetValue(_newReference_).
+          1. If the code matched by this production is strict mode code, let _strict_ be *true*, else let _strict_ be *false*.
+          1. Return ? EvaluateDynamicPropertyAccess(_newValue_, |Expression|, _strict_).
+        </emu-alg>
+        <emu-grammar>OptionalChain : OptionalChain `.` IdentifierName</emu-grammar>
+        <emu-alg>
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Let _newReference_ be ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
+          1. Let _newValue_ be ? GetValue(_newReference_).
```
Comment (bakkot) [inline spec.html]: Should there not be a

```
          1. If _newValue_ is *undefined* or *null*, then
            1. Return *undefined*.
```
after this line? (Ditto the ChainEvaluation sections for ``OptionalChain `[` Expression `]`` and ``OptionalChain : OptionalChain Arguments``.)

File: spec.html (hunk 3 above)
Comment (rkirsling) [reply spec.html]: This was fixed in https://github.com/tc39/proposal-optional-chaining/pull/120 but must have not been propagated to this PR.

File: spec.html (hunk 4 above)
Comment (rkirsling) [reply spec.html]: No, if we have `a?.b.c`, then the chain is `?.b.c`—_whether_ we evaluate a chain is conditional but the evaluation of the chain is not.

Comment (bakkot) [reply spec.html]: Oh, thank you, yes. If a `return` were necessary it would have occurred before this algorithm were reached.

File: spec.html (hunk 1 above)
Comment (claudepache) [reply spec.html]: > it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records,

That was reverted to the traditional form “ChainEvaluation of ... with arguments ...” in tc39/proposal-optional-chaining#118, but te change has not been forwarded to this PR.

**Comment by syg** [comment]
Could someone update this PR to reflect the upstream fixes and editorial changes?

File: spec.html [hunk 5]
```diff
@@ -13569,29 +13639,52 @@ <h1>Runtime Semantics: Evaluation</h1>
...
+          1. Let _baseReference_ be the result of evaluating |CallExpression|.
+          1. Let _baseValue_ be ? GetValue(_baseReference_).
+          1. If the code matched by this |CallExpression| is strict mode code, let _strict_ be *true*, else let _strict_ be *false*.
+          1. Return ? EvaluateStaticPropertyAccess(_baseValue_, |IdentifierName|, _strict_);
+        </emu-alg>
       </emu-clause>
     </emu-clause>
 
+    <emu-clause id="sec-evaluate-dynamic-property-access" aoid="EvaluateDynamicPropertyAccess">
+      <h1>Runtime Semantics: EvaluateDynamicPropertyAccess(_baseValue_, _expression_, _strict_ )</h1>
```
Comment (syg) [inline spec.html]: Hm... I'm not a fan of the Static/Dynamic naming here. Maybe ExpressionPropertyAccess and IdentifierPropertyAccess? @bakkot @ljharb wdyt?

Comment (michaelficarra) [reply spec.html]: "static" and "computed" are the terms I typically see/use

Comment (syg) [reply spec.html]: I'm fine with that too, though I still prefer the more literal names -- doesn't leave room for misinterpreting what exactly is static or computed.

Comment (DanielRosenwasser) [reply spec.html]: What about `EvaluateExpressionKeyPropertyAccess` and `EvaluateIdentifierKeyPropertyAccess`?

**Comment by DanielRosenwasser** [comment]
I've integrated the upstream changes and addressed feedback provided apart from the `Evaluate_______PropertyAccess` naming discussion which I'll try to address shortly

File: spec.html [hunk 6]
```diff
@@ -13379,9 +13387,25 @@ <h2>Syntax</h2>
...
+
+      OptionalChain[Yield, Await] :
+        `?.` `[` Expression[+In, ?Yield, ?Await] `]`
+        `?.` IdentifierName
+        `?.` Arguments[?Yield, ?Await]
+        `?.` TemplateLiteral[?Yield, ?Await, +Tagged]
+        OptionalChain[?Yield, ?Await]  `[` Expression[+In, ?Yield, ?Await] `]`
+        OptionalChain[?Yield, ?Await] `.` IdentifierName
+        OptionalChain[?Yield, ?Await] Arguments[?Yield, ?Await]
+        OptionalChain[?Yield, ?Await] TemplateLiteral[?Yield, ?Await, +Tagged]
```
Comment (bakkot) [inline spec.html]: I would be happier if these were in the same order as they are in [CallExpression](https://tc39.es/ecma262/#prod-CallExpression), i.e.

```
      OptionalChain[Yield, Await] :
        `?.` Arguments[?Yield, ?Await]
        `?.` `[` Expression[+In, ?Yield, ?Await] `]`
        `?.` IdentifierName
        `?.` TemplateLiteral[?Yield, ?Await, +Tagged]
        OptionalChain[?Yield, ?Await] Arguments[?Yield, ?Await]
        OptionalChain[?Yield, ?Await]  `[` Expression[+In, ?Yield, ?Await] `]`
        OptionalChain[?Yield, ?Await] `.` IdentifierName
        OptionalChain[?Yield, ?Await] TemplateLiteral[?Yield, ?Await, +Tagged]
```

File: spec.html [hunk 7]
```diff
@@ -13393,6 +13417,33 @@ <h2>Supplemental Syntax</h2>
...
+            `?.` TemplateLiteral[?Yield, ?Await, +Tagged]
+            OptionalChain[?Yield, ?Await] TemplateLiteral[?Yield, ?Await, +Tagged]
+        </emu-grammar>
+        <ul>
+          <li>
+            It is a Syntax Error if any code matches this production.
+          </li>
+        </ul>
+        <emu-note>
+          <p>This production exists in order to prevent automatic semicolon insertion rules (<emu-xref href="#sec-automatic-semicolon-insertion"></emu-xref>) to be applied to the following code:</p>
```
Comment (bakkot) [inline spec.html]:
```suggestion
          <p>This production exists in order to prevent automatic semicolon insertion rules (<emu-xref href="#sec-automatic-semicolon-insertion"></emu-xref>) from applying to the following code:</p>
```

File: spec.html [hunk 8]
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+            CallExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be |CallExpression|.
+          1. Let _baseReference_ be the result of evaluating _baseExpression_.
+          1. Let _baseValue_ be ? GetValue(_baseReference_).
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
```
Comment (bakkot) [inline spec.html]:
```suggestion
          1. Return the result of performing ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_.
```

File: spec.html [hunk 9]
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+            OptionalExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be |OptionalExpression|.
+          1. Let _baseReference_ be the result of evaluating _baseExpression_.
+          1. Let _baseValue_ be ? GetValue(_baseReference_).
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
```
Comment (bakkot) [inline spec.html]:
```suggestion
          1. Return the result of performing ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_.
```

File: spec.html (hunk 5 above)
Comment (michaelficarra) [reply spec.html]: I don't like `Identifier` in the name at all. It's an inline string-named property access. And computed property accesses are property accesses where the property name is computed by evaluating a given expression. So `ExpressionPropertyAccess` / `ComputedPropertyAccess` and `StringPropertyAccess` / `StaticPropertyAccess` / `FixedPropertyAccess` are best IMO.

I also don't like `Key` in the name unless it's preceded by `Property`.

File: spec.html [hunk 10]
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+      <h1>Optional Chains</h1>
+      <emu-note>An optional chain is a chain of one or more property accesses and function calls, the first of which begins with the token `?.`.</emu-note>
+      <emu-clause id="sec-optional-chaining-evaluation">
+        <h1>Runtime Semantics: Evaluation</h1>
+        <emu-grammar>
+          OptionalExpression :
+            MemberExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be |MemberExpression|.
```
Comment (bakkot) [inline spec.html]: I'm not sure the `_baseExpression_` and `_optionalChain_` aliases add any clarity in these algorithms. Would it not be better to just have

```
1. Let _baseReference_ be the result of evaluating |MemberExpression|.
```
and similarly
```
1. Return the result of performing ChainEvaluation of |OptionalChain| with arguments _baseValue_ and _baseReference_.
```
thereby avoiding a layer of indirection? That's what the rest of the spec tends to do.

File: spec.html [hunk 11]
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
...
+        <emu-grammar>OptionalChain : `?.` Arguments</emu-grammar>
+        <emu-alg>
+          1. Let _thisChain_ be this |OptionalChain|.
+          1. Let _tailCall_ be IsInTailPosition(_thisChain_).
+          1. Return ? EvaluateCall(_baseValue_, _baseReference_, |Arguments|, _tailCall_).
+        </emu-alg>
+        <emu-grammar>OptionalChain : OptionalChain `[` Expression `]`</emu-grammar>
+        <emu-alg>
+          1. Let _optionalChain_ be |OptionalChain|.
+          1. Let _newReference_ be ? ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_).
```
Comment (bakkot) [inline spec.html]:
```suggestion
          1. Let _newReference_ be ? ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_.
```

File: spec.html [hunk 12]
```diff
@@ -13428,6 +13479,19 @@ <h1>Static Semantics: Contains</h1>
           1. If _symbol_ is an |Identifier| and StringValue of _symbol_ is the same value as the StringValue of |IdentifierName|, return *true*.
           1. Return *false*.
         </emu-alg>
+        <emu-grammar>OptionalChain : `?.` IdentifierName</emu-grammar>
+        <emu-alg>
+          1. If _symbol_ is a |ReservedWord|, return *false*.
+          1. If _symbol_ is an |Identifier| and StringValue of _symbol_ is the same value as the StringValue of |IdentifierName|, return *true*.
```
Comment (bakkot) [inline spec.html]: If https://github.com/tc39/ecma262/pull/1519 lands before this PR, this step will need to be removed both places it occurs.

File: spec.html (hunk 10 above)
Comment (claudepache) [reply spec.html]: > I'm not sure the _baseExpression_ and _optionalChain_ aliases add any clarity in these algorithms. 

It is an undue leftover of the original spec, [that said](https://tc39.es/proposal-optional-chaining/#sec-optional-chaining-evaluation): “Let _baseExpression_ be the first child of this _OptionalExpression_ (i.e., _MemberExpression_, _CallExpression_, or _OptionalExpression_)”. Now that the algorithm is repeated thrice (once per possible first child of _OptionalExpression_), there is indeed no point in having _baseExpression_.

File: spec.html (hunk 5 above)
Comment (claudepache) [reply spec.html]: Another option would be `EvaluateDotPropertyAccess` and `EvaluateBracketPropertyAccess` (although I’m not exactly a fan of that).

Note that, with the advent of private fields and methods, there may be a future `EvaluatePrivatePropertyAccess`.

Comment (DanielRosenwasser) [reply spec.html]: > Note that, with the advent of private fields and methods, there may be a future `EvaluatePrivatePropertyAccess`.

In which case I think that `Identifier` and `Expression` are probably the best differentiators - because then it can simply be named `EvaluatePrivateNamePropertyAccess` or `EvaluatePrivateNameKeyPropertyAccess`

Comment (DanielRosenwasser) [reply spec.html]: I've gone with the naming I mentioned above, since it feels like a good way to leave the door open to *PrivateName` keys, but I didn't use the `PropertyKey` naming scheme because `PropertyKeyPropertyAccess` gets a bit unwieldy. Long term, if there's an issue, I figure it can always be renamed.

File: spec.html [hunk 13]
```diff
@@ -13584,29 +13654,52 @@ <h1>Runtime Semantics: Evaluation</h1>
...
+        1. Let _propertyNameValue_ be ? GetValue(_propertyNameReference_).
+        1. Let _bv_ be ? RequireObjectCoercible(_baseValue_).
+        1. Let _propertyKey_ be ? ToPropertyKey(_propertyNameValue_).
+        1. Return a value of type Reference whose base value component is _bv_, whose referenced name component is _propertyKey_, and whose strict reference flag is _strict_.
+    </emu-clause>
+    <emu-clause id="sec-evaluate-identifier-key-property-access" aoid="EvaluateIdentifierKeyPropertyAccess">
+      <h1>Runtime Semantics: EvaluateIdentifierKeyPropertyAccess( _baseValue_, _identifierName_, _strict_ )</h1>
+      <p>The abstract operation EvaluateIdentifierKeyPropertyAccess takes as arguments a value _baseValue_, a Parse Node _identifierName_, and a Boolean argument _strict_. It performs the following steps:</p>
+      <emu-alg>
+        1. Let _bv_ be ? RequireObjectCoercible(_baseValue_).
```
Comment (bakkot) [inline spec.html]: I would add
```
1. Assert: _identifierName_ is a |IdentifierName|
```
before this line, to justify the use of StringValue below.

**Comment by rkirsling** [comment]
Hmm, `Expression` / `Identifier` is one thing, but I don't think I understand why `Key` is in there.

**Comment by DanielRosenwasser** [comment]
@rkirsling it was just to disambiguate that the *Identifier* or *Expression* is the key, not the base value.

**Comment by rkirsling** [comment]
Not sure how the others feel, but that suggestion sounds pretty good to me.

**Comment by michaelficarra** [comment]
I really hate bikeshedding something as unimportant as this, but really, "Identifier"/"IdentifierName" should not be in these names.

**Comment by syg** [comment]
I am fine with literally anything for those names so long as it doesn't require me to intuit what's dynamic and what's static.

**Comment by devsnek** [comment]
In engine262 we'd call them `EvaluateMemberExpressionWithIdentifier` and `EvaluateMemberExpressionWithExpression`

**Comment by DanielRosenwasser** [comment]
I think using the syntactically-guided names (as it currently stands) is probably the best way forward. I'm also fine with switching to `EvaluatePropertyAccessWith[Identifier|Expression]Key` if there's a preference there.

Keep in mind that this can be switched later if it causes legitimate issues, so I'd urge people not to keep picking out new colors for the bikeshed. 😄

**Comment by rkirsling** [comment]
I'm fine with virtually everything that's been suggested (particularly the recent suggestions involving `By` / `With`), but I do think the current naming in particular is hard to parse. It makes me ask, "What's a _key property_? Or is it a _key property access_?"

File: spec.html [hunk 14]
```diff
@@ -21710,6 +21890,34 @@ <h1>Expression Rules</h1>
...
+          OptionalChain :
+            `?.` `[` Expression `]`
+            `?.` IdentifierName
+            OptionalChain `[` Expression `]`
+            OptionalChain `.` IdentifierName
+        </emu-grammar>
+        <emu-alg>
+          1. Return *false*.
+        </emu-alg>
+        <emu-grammar>
```
Comment (rkirsling) [inline spec.html]: A final nitpick: Given that the [ordering of these productions was changed](https://github.com/tc39/ecma262/pull/1646#discussion_r345970102), we should probably put the Arguments case before the Expression and IdentifierName cases here too.

Comment (rkirsling) [reply spec.html]: (Then again, I guess the Member-Call-Optional ordering isn't being preserved here either...hmm.)
